source/extensions/dynamic_modules/sdk/rust/src/lib.rs

```diff
@@ -95,13 +95,95 @@ pub static NEW_HTTP_FILTER_CONFIG_FUNCTION: OnceLock<NewHttpFilterConfigFunction
 /// imlementation is recommended to implement the [`Drop`] trait to handle the necessary cleanup.
 pub trait HttpFilterConfig {
   /// This is called when a HTTP filter chain is created for a new stream.
-  fn new_http_filter(&self) -> Box<dyn HttpFilter> {
-    unimplemented!() // TODO.
+  fn new_http_filter(&self, _envoy: EnvoyHttpFilterConfig) -> Box<dyn HttpFilter> {
+    panic!("not implemented");
   }
 }
 
-/// The trait that represents an Envoy Http filter for each stream.
-pub trait HttpFilter {} // TODO.
+/// The trait that corresponds to an Envoy Http filter for each stream
+/// created via the [`HttpFilterConfig::new_http_filter`] method.
+pub trait HttpFilter {
+  /// This is called when the request headers are received.
+  /// The `envoy_filter` can be used to interact with the underlying Envoy filter object.
+  /// The `end_of_stream` indicates whether the request is the last message in the stream.
+  ///
+  /// This must return [`abi::envoy_dynamic_module_type_on_http_filter_request_headers_status`] to
+  /// indicate the status of the request headers processing.
+  fn on_request_headers(
+    &mut self,
+    _envoy_filter: EnvoyHttpFilter,
+    _end_of_stream: bool,
+  ) -> abi::envoy_dynamic_module_type_on_http_filter_request_headers_status {
+    abi::envoy_dynamic_module_type_on_http_filter_request_headers_status::Continue
+  }
```

Comment:
I'm not suggesting to tackle this now as it's a huge project, but I think technically it's possible that this could be an async trait, and an async executor could be implemented directly on top of the Envoy event loop. That would be pretty neat. As it stands if any async work has to happen the only way to do it is going to be to move it off to a thread pool I think and communicate back and forth with the Envoy threads, or I guess expose all of the Envoy async callback stuff directly. It would be so nice if the Rust filters were async and worked more like how Lua works. Again I believe this is technically possible so something to think about for the future.

Reply:
ack, and in fact this is something that came up in our conversation with Linkedin's team (cc @thenewwazoo). I think it's going to require non-negligible amount of changes to Envoy code, but agree pretty neat and potentially improve performance (i guess?).

Reply:
👋 I'm actually not working on this at the moment, but I bet @tylerhorth would be interested! :)